keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,556 @@
+authorizers:
+- type: Webhook
+  webhook:
+    server: https://<path-to-webserver>
+    authentication:
+        certificateAuthority: </path/to/ca.pem> # either this or certificateAuthorityData
+        clientCertificate: </path/to/cert.pem> # either this or clientCertificateData
+        clientKey: </path/to/key.pem> # either this or clientKeyData
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: Deny | NoOpinion
+- type: Node
+- type: RBAC
+- type: Webhook
+  webhook:
+    kubeConfigFile: <path-to-allow-webhook-config>
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: NoOpinion
+```
+
+Validation will allow multiple authorizers of type "Webhook" to be added to the config, but one authorizer each for other types.
+The ordering of this chain will be decided by the order specified in the file.
+
+The keys `kubeConfigFile`, `authorizedTTL`, `unauthorizedTTL` and `subjectAccessReviewVersion` version accept values corresponding to flags `--authorization-webhook-config-file`, `--authorization-webhook-cache-authorized-ttl`, `--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version` respectively.
+
+This diagram shows how the configuration flags map to configuration file keys for a webhook:
+![](https://i.imgur.com/PJSmgOq.jpg)
+
+The `onError` will be a required field which allow users to specify whether or not request is denied if the webhook errors out or is unreachable.
+This allows `SubjectAccessReviews` to be more definitive.
+
+Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding flag is not supplied.
+While configuring authorization modes using the file config, the version supported by a webhook has to be mentioned using a required field `subjectAccessReviewVersion`.
+
+Authentication of webhooks will move to the configuration file and would be mapped as shown in the diagram above.
+
+> Note: ABAC will not be supported by the new file format as we discourage the usage of this mode.
+
+The code path for enabling the above will only be triggered if the feature flag will be enabled until the time the feature flag is removed and configuring authorizer through a file becomes GA.
+
+> TODO: More details on the implementation will be charted once the summary, motivation, goals, non-goal and proposal are approved.
+
+### Monitoring
+
+We will add the following 4 metrics:
+
+1. `apiserver_authorization_step_invocations_total`
+
+This will be incremented on round-trip of an authorizer. It will track total authorization decision invocations across the following labels.
+
+- `mode` {RBAC, Node, Webhook}
+- `decision` {Allow, Deny, NoOpinion}
+
+2. `apiserver_authorization_step_webhook_invocations_total`
+
+This will be incremented on round-trip of an authorization webhook. It will track total invocation counts across the following labels.
+
+- `server`
+- `code` {2xx, 4xx, 5xx}
```

Comment:
Do these result in high cardinality metrics?

Reply:
The cardinality should be 3. The only labels maintained would be `2xx` catching all 200 series response codes and so on.

Reply:
Maybe not `code` then but `responseCodeGroup` (`2xx` is a group of codes, not a code itself).

Reply:
Where does a 300 response appear in the `apiserver_authorization_step_webhook_invocations_total` time series group?

Reply:
> Maybe not code then but responseCodeGroup (2xx is a group of codes, not a code itself).

Thanks for the idea. If we end up keeping this label, I will go with this.